src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,315 @@
+
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.SyncNamespace
+{
+    [ExportLanguageService(typeof(ISyncNamespaceService), LanguageNames.CSharp), Shared]
+    internal sealed class CSharpSyncNamespaceService :
+        AbstractSyncNamespaceService<NamespaceDeclarationSyntax, CompilationUnitSyntax, MemberDeclarationSyntax>
+    {
+        protected override SyntaxList<MemberDeclarationSyntax> GetMemberDeclarationsInContainer(SyntaxNode compilationUnitOrNamespaceDecl)
+        {
+            if (compilationUnitOrNamespaceDecl is NamespaceDeclarationSyntax namespaceDecl)
+            {
+                return namespaceDecl.Members;
+            }
+            else if (compilationUnitOrNamespaceDecl is CompilationUnitSyntax compilationUnit)
+            {
+                return compilationUnit.Members;
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        /// <summary>
+        /// Try to get a new node to replace given node, which is a reference to a top-level type declared inside the namespce to be changed.
+        /// If this reference is the right side of a qualified name, the new node returned would be the entire qualified name. Depends on 
+        /// whether <paramref name="newNamespaceParts"/> is provided, the name in the new node might be qualified with this new namespace instead.
+        /// </summary>
+        /// <param name="reference">A reference to a type declared inside the namespce to be changed, which is calculated based on results from 
+        /// `SymbolFinder.FindReferencesAsync`.</param>
+        /// <param name="newNamespaceParts">If specified, and the reference is qualified with namespace, the namespace part of original reference 
+        /// will be replaced with given namespace in the new node.</param>
+        /// <param name="old">The node to be replaced. This might be an ancestor of original reference.</param>
+        /// <param name="new">The replacement node.</param>
+        public override bool TryGetReplacementReferenceSyntax(
+            SyntaxNode reference,
+            ImmutableArray<string> newNamespaceParts, 
+            ISyntaxFactsService syntaxFacts, 
+            out SyntaxNode old, 
+            out SyntaxNode @new)
+        {
+            if (!(reference is SimpleNameSyntax nameRef))
+            {
+                old = @new = null;
+                return false;
```

Comment:
aside, do we have tests that validate we fix up `<see cref...>` tags in doc-comments properly?

Reply:
That's a good point, I never thought about those scenarios. It probably doesn't because when I try to get the syntax node from the reference finder, we don't dive into trivia.

Reply:
Added support and tests for cref.